nova/virt/vmwareapi/vmops.py

```diff
@@ -2058,13 +2137,87 @@ def migrate_disk_and_power_off(self, context, instance, dest, flavor):
                                        total_steps=RESIZE_TOTAL_STEPS)
 
         # 1. Power off the instance
-        vm_util.power_off_instance(self._session, instance, vm_ref)
         self._update_instance_progress(context, instance,
                                        step=1,
                                        total_steps=RESIZE_TOTAL_STEPS)
+        vm_util.power_off_instance(self._session, instance, vm_ref)
+
+        # 2. Detach the volumes
+        self._update_instance_progress(context, instance,
+                                       step=2,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        self._detach_volumes(instance, block_device_info)
 
-    def confirm_migration(self, migration, instance, network_info):
+        # 3. Detach network interfaces
+        self._update_instance_progress(context, instance,
+                                       step=3,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        try:
+            device_change = self._get_network_device_change(
+                vm_ref, instance.image_meta, network_info, operation="remove")
+            vm_util.vm_device_change(self._session, vm_ref, device_change)
+        except Exception as error:
+            # If the network detach fails, we attempt to rollback the volume
+            # attachments and power on, then keep the VM in an ACTIVE state
+            adapter_type = vm_util.get_adapter_type(self._session, instance,
+                                                    vm_ref)
+            self._attach_volumes(instance, block_device_info, adapter_type,
+                                 vm_ref)
+            vm_util.power_on_instance(self._session, instance, vm_ref)
```

Comment:
We should not unconditionally power-on the instance. Can we somehow get the state the instance was in before we started the migration effort? I know that users like to power off their instances before they resize.

Reply:
Done.